### ISSUE TITLE
Better names for onNextDropped vs doOnDiscard hooks, compose/transform, doOnEach

### ISSUE THREAD

**Issue by thekalinga** [reactor/reactor-core#1381]
Currently the hooks for handling dropped elements due to reactive spec breakage vs normal flow  are very similarly sounding. Unless one looks at the documentation, its difficult to determine which one is which

i.e its difficult to distinguish between `onNextDropped` vs `doOnDiscard` on whether they apply in one scenario vs another. Its better to have more easy to remember, yet obvious names than what we currently have

PS: As of now, I am not sure of all the uses of these two hooks to comeup with better names

**Comment by smaldini** [comment]
onNextDropped are unexpected, mostly protocol failures. I agree there is an issue here to look at, where you looking at these concepts in StepVerifier ?

**Comment by thekalinga** [comment]
@smaldini In the documentation

**Comment by thekalinga** [comment]
@smaldini I think we need better names for `compose` & `transform` aswell.

I have few suggestions

`compose` -> `composeOnSubscribe`
`transform` -> `composeAtAssembly`

What do you say?

**Comment by thekalinga** [comment]
Also `doOnEach` -> `doOnSignal`. The second is more explicit & does not require user to lookup documentation, while the former does require documentation lookup

**Comment by simonbasle** [comment]
We cannot easily change public API method names like that, although I can see your point.

The `doOnDiscard` comment is particularly unfortunate since it was only released recently in 3.2.0 (but `doOnError` has been there for a long time). I'm not entirely convinced this is more than a minor inconvenience for the hooks names though, because hooks shouldn't be used purely on what the developer think their names mean => advanced feature that preferably needs a good read of the documentation anyways.

But `compose` vs `transform` and `doOnEach` I'd like to change when possible, **which is not before at least 3.4**...

**Comment by thekalinga** [comment]
@simonbasle Thanks for the clarification. I only started going over the documentation to understand various aspects and attempting to map them to source. So, I came across `doOnDiscard` only now :(

>  I'm not entirely convinced this is more than a minor inconvenience for the hooks names though, because hooks shouldn't be used purely on what the developer think their names mean => advanced feature that preferably needs a good read of the documentation anyways.

I agree that developers needs read documentation when using an advanced feature. But the problem is, the same developer needs to read documentation everytime to make sure that he is not using the wrong one (just like `compose` vs `transform`) almost every single time he needs to hook into the system, as both the names are very similar sounding
